### PR TITLE
Say hello to the new Simperby architecture

### LINKED ISSUE #416: Include `round` in the previous block finalization proof
Precommits must be signed on data which cryptographically contains information about the consensus round.
Currently precommits are signed only on the block hash, which does not contain any information about the round.
This will make it possible to replay commits in different rounds.
Thus `BlockHeader::prev_block_finalization_proof` must be signed on `(BlockHeader, Round)`, not just `BlockHeader`

[reference (Commit::Round)](https://github.com/tendermint/spec/blob/8dd2ed4c6fe12459edeb9b783bdaaaeb590ec15c/spec/core/data_structures.md#commit)

### PR DESCRIPTION
**Pull request by junha1**
## What's changed (notable only)

- Finalization proof now contains information about round.
- Simperby node works under an existing repository
- Local works (commits not made by simperby; e.g., remote tracking branches, manually created commit) can be synced
- Brand new DMS with better interface
- Consensus module entirely refactored
- Repository module entirely refactored
- `Sync` command removed

## What's next

- Implement node
- Implement integration tests for consensus/node
- #417 #415 #411 #407 #394 
- DMS interface revisit
- Agenda/Finalization proof revisit (subject to change)

## Fixes

fixes #416 #372 #382 #385 

## Reviews

Please check commit messages (though not that friendly)

- @hyeonLewis e7e5aa9
- @TomTaehoonKim e7e5aa9 43d8748 e2e8d91 73f7a88 and others if you want
- @lemonholic e7e5aa9 a7887f8 4beb857 705236b
- @YoonhoKim0527 e7e5aa9 4beb857 43d8748
- @JeongHunP fc63153 e7e5aa9 43d8748

### CODE REVIEW

File: repository/src/raw/mod.rs [hunk 1]
```diff
@@ -256,17 +254,22 @@ impl RawRepository {
...
         helper_0_mut(self, RawRepositoryInner::stash).await
     }
 
-    /// Apply the most recent stash.
+    /// Pops the most recent stash.
+    pub async fn stash_pop(&mut self) -> Result<(), Error> {
+        helper_0_mut(self, RawRepositoryInner::stash_pop).await
+    }
+
+    /// Applys the most recent stash.
```
Comment (JeongHunP) [inline repository/src/raw/mod.rs:267]:
```suggestion
    /// Applies the most recent stash.
```

File: repository/src/lib.rs [hunk 2]
```diff
@@ -357,210 +181,74 @@ impl DistributedRepository {
...
     /// 5. the reserved state in a valid format.
     pub async fn check(&self, _starting_height: BlockHeight) -> Result<bool, Error> {
-        unimplemented!()
+        todo!()
     }
 
-    /// Synchronizes the `finalized` branch to the given commit.
+    // ---------------
+    // Operations that interacts with possible local works
+    // (manually added commits or remote tracking branchess)
```
Comment (JeongHunP) [inline repository/src/lib.rs:189]:
```suggestion
    // (manually added commits or remote tracking branches)
```

File: repository/src/lib.rs [hunk 3]
```diff
@@ -357,210 +181,74 @@ impl DistributedRepository {
     /// 4. the canonical history of the `finalized` branch.
     /// 5. the reserved state in a valid format.
     pub async fn check(&self, _starting_height: BlockHeight) -> Result<bool, Error> {
-        unimplemented!()
+        todo!()
     }
 
-    /// Synchronizes the `finalized` branch to the given commit.
+    // ---------------
+    // Operations that interacts with possible local works
```
Comment (JeongHunP) [inline repository/src/lib.rs:188]:
```suggestion
    // Operations that interact with possible local works
```

File: repository/tests/integration_test.rs [hunk 4]
```diff
@@ -1,25 +1,26 @@
...
-
     let server_node_dir = create_temp_dir();
     setup_pre_genesis_repository(&server_node_dir, rs.clone()).await;
 
     let mut server_node_repo = DistributedRepository::new(
-        RawRepository::open(&format!("{server_node_dir}/repository/repo"))
-            .await
-            .unwrap(),
+        Arc::new(RwLock::new(
+            RawRepository::open(&format!("{server_node_dir}/repository"))
```
Comment (JeongHunP) [inline repository/tests/integration_test.rs:20]: Why did you remove `/repo`?

Comment (junha1) [reply repository/tests/integration_test.rs:20]: It's the result of #385.

**Comment by TomTaehoonKim** [comment]
Typo in message of https://github.com/postech-dao/simperby/commit/43d8748c31253be9690e4ba8e3d5a8bce53be71f.
multipe > multiple.

File: repository/src/interpret/read.rs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+/// It does not accept a last finalization proof commit.
+///
+/// - Returns `Ok(Ok(commit))` if the branch (taking the given commit as a tip) is valid.
+/// - Returns `Ok(Err(error))` if the branch is invalid.
+/// - Returns `Err(error)` if there was an error reading the commits.
+pub async fn read_and_verify_commits_from_last_finalized_block(
+    raw: &RawRepository,
+    commit_hash: CommitHash,
+) -> Result<Result<CommitSequenceVerifier, Error>, Error> {
+    let lfi = read_last_finalization_info(raw).await?;
```
Comment (TomTaehoonKim) [inline repository/src/interpret/read.rs:136]: Is `last_finalization_info` just too long for a variable name?

File: network/src/dms/messages.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
+use super::*;
+use serde::{de::DeserializeOwned, ser::Serialize};
+
+pub type DmsKey = String;
+
+pub trait DmsMessage: Send + Sync + 'static + ToHash256 + Serialize + DeserializeOwned {
+    /// Checks if the message is valid.
+    fn check(&self) -> Result<(), Error>;
+
+    /// Defines how to commit a message, by cryptographicically signing it.
```
Comment (lemonholic) [inline network/src/dms/messages.rs:10]:
```suggestion
    /// Defines how to commit a message, by cryptographically signing it.
```

File: network/src/dms/messages.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    ///
+    /// In case that the message can't be guaranteed to be unique among other protocols,
+    /// this method provides `dms_key` to be used as a unique identifier.
+    ///
+    /// One potential use case other than the default implemenations which ignores `dms_key`
```
Comment (lemonholic) [inline network/src/dms/messages.rs:15]:
```suggestion
    /// One potential use case other than the default implementation which ignores `dms_key`
```

File: repository/src/interpret/works.rs [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                commits[index].1,
+                LastFinalizationProof {
+                    height: second_to_last_header.height,
+                    proof: last_header.prev_block_finalization_proof.clone(),
+                },
+            )
+            .await?;
+        }
+
+        // Create the associate branch
```
Comment (TomTaehoonKim) [inline repository/src/interpret/works.rs:128]:
```suggestion
        // Create a branch associated to the last commit
```
It seems as if there is a term "associate branch" so how about fixing it like this?

File: repository/src/interpret/works.rs [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        .await?
+        .into_iter()
+        .filter(|s| {
+            s.as_str() != WORK_BRANCH_NAME
+                && s.as_str() != FINALIZED_BRANCH_NAME
+                && s.as_str() != FP_BRANCH_NAME
+                && !s.starts_with("a-")
+                && !s.starts_with("b-")
+        })
+        .collect();
```
Comment (TomTaehoonKim) [inline repository/src/interpret/works.rs:174]: Was only leaving local p branches your intention?

File: repository/src/interpret/works.rs [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+    // Note that remote branches are automatically removed when the remote repository is removed.
+    let remote_list = raw.list_remotes().await?;
+    for (remote_name, _) in remote_list {
+        raw.remove_remote(remote_name).await?;
+    }
+
+    Ok(())
+}
+
+pub async fn sync_old(
```
Comment (TomTaehoonKim) [inline repository/src/interpret/works.rs:256]: Interface for `sync_old` is missing. Please add it with comments.

File: cli/src/cli.rs [hunk 11]
```diff
@@ -76,28 +84,13 @@ pub enum Commands {
     /// Clone a remote Simperby repository to the current directory,
     /// and initialize a new Simperby node after verification.
     ///
-    /// This is same as `git clone && simperby init`.
+    /// This is same as `git clone && cd <directory> && simperby init`.
```
Comment (TomTaehoonKim) [inline cli/src/cli.rs:87]: What is <directory> here?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: network/src/dms/messages.rs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    /// is when the messages of which the signature is presented to another protocol,
+    /// so not aggregating `dms_key`, (which is a very specific implementation detail of of DMS)
+    /// makes sense.
```
[the remark this answers is not in the document]
Comment (lemonholic) [reply network/src/dms/messages.rs:18]:
```suggestion
    /// Note that `dms_key` is ignored in the default implementation.
    /// One potential use case other than the default is presenting messages with the signature to another protocol.
    /// Such case leverages a very specific implementation detail, so it makes sense not to aggregate `dms_key` by default.
```
I'm totally not sure if I understood it correctly.

File: cli/src/cli.rs (hunk 11 above)
Comment (junha1) [reply cli/src/cli.rs:87]: You need to go into the cloned directory.

File: repository/src/interpret/works.rs (hunk 10 above)
Comment (junha1) [reply repository/src/interpret/works.rs:256]: It's legacy. Will be removed soon.

File: network/src/dms/messages.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (junha1) [reply network/src/dms/messages.rs]: You understood correctly, but I'm not sure if this tricky interface worth maintaining. Let me revisit this on the upcoming PR.

File: repository/tests/integration_test.rs [hunk 13]
```diff
@@ -35,19 +36,21 @@ async fn basic_1() {
 
     let client_node_dir = create_temp_dir();
     simperby_test_suite::run_command(format!(
-        "cd {client_node_dir} && mkdir repository && cd repository && cp -r {server_node_dir}/repository/repo {client_node_dir}/repository"
+        " cp -r {server_node_dir}/repository {client_node_dir}/repository"
```
[the remark this answers is not in the document]
Comment (junha1) [reply repository/tests/integration_test.rs:39]: fixed

File: network/src/dms/mod.rs [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+                .add_or_overwrite_file(
+                    &format!("message-{message_hash}.json"),
+                    serde_spb::to_string(&message).unwrap(),
+                )
+                .await?;
+        };
+        Ok(())
+    }
+
+    async fn retrieve_packtes(&self) -> Result<Vec<Packet>, Error> {
```
Comment (lemonholic) [inline network/src/dms/mod.rs:326]: Need to be refactored!
(packtes -> packets)

File: repository/src/interpret/dms.rs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+pub async fn flush(
+    _raw: Arc<RwLock<RawRepository>>,
+    _dms: Arc<RwLock<Dms<Message>>>,
+) -> Result<(), Error> {
+    todo!()
+}
+
+/// Updates the repository module with the latest messages from the DMS.
```
[the remark this answers is not in the document]
Comment (junha1) [reply repository/src/interpret/dms.rs:53]: It does.

File: cli/src/cli.rs [hunk 16]
```diff
@@ -42,6 +42,14 @@ pub enum CreateCommands {
     Agenda,
 }
 
+#[derive(Debug, Subcommand)]
+pub enum PeerCommand {
+    Add { address: String },
+    Remove { name: String },
```
Comment (lemonholic) [inline cli/src/cli.rs:48]: How about allowing optional `name` for the `Add` command?

File: cli/src/cli.rs [hunk 17]
```diff
@@ -42,6 +42,14 @@ pub enum CreateCommands {
...
+    Update,
```
Comment (lemonholic) [inline cli/src/cli.rs:49]: What is this command for?

File: repository/src/interpret/works.rs (hunk 9 above)
Comment (junha1) [reply repository/src/interpret/works.rs:174]: Nope mistake. fixed.

File: network/src/dms/messages.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lemonholic) [reply network/src/dms/messages.rs]: Okay.

File: common/src/types.rs [hunk 18]
```diff
@@ -31,6 +30,27 @@ pub struct Member {
...
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct FinalizationSignTarget {
+    pub block_hash: Hash256,
+    pub round: ConsensusRound,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct FinalizationProof {
+    pub round: ConsensusRound,
```
Comment (hyeonLewis) [inline common/src/types.rs:41]: What does `ConsensusRound` mean? And should the treasury contract hold this as storage?

File: repository/src/interpret/read.rs (hunk 5 above)
Comment (junha1) [reply repository/src/interpret/read.rs:136]: I think so. Like `fp` or `dms`. I think it's worth having a doc for those abbreviations

File: common/src/types.rs (hunk 18 above)
Comment (junha1) [reply common/src/types.rs:41]: In short, `ConsensusRound` is a value that is decided during the consensus process.
All you need to do is to understand the type of `FinalizationProof` has been changed, which is the type of `BlockHeader::prev_block_finalization_proof` and `proof` in `update_treasury_light_client`.

File: cli/src/cli.rs (hunk 17 above)
Comment (junha1) [reply cli/src/cli.rs:49]: Synchronous fetch/broadcast of peer discovery protocol. This may add some remote repositories

File: network/src/dms/mod.rs (hunk 14 above)
Comment (junha1) [reply network/src/dms/mod.rs:326]: fixed.

File: cli/src/cli.rs (hunk 16 above)
Comment (junha1) [reply cli/src/cli.rs:48]: fixed

File: repository/src/interpret/works.rs (hunk 9 above)
Comment (junha1) [reply repository/src/interpret/works.rs:174]: Btw I believe `p` is more appropriate as a tag. Let's revisit this later

**Review by JeongHunP — APPROVED** [review]

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by YoonhoKim0527 — APPROVED** [review]

File: cli/src/cli.rs (hunk 17 above)
Comment (lemonholic) [reply cli/src/cli.rs:53]: ok

**Review by lemonholic — APPROVED** [review]

**Review by hyeonLewis — APPROVED** [review]
